pkg/mockclient/mockclient.go

```diff
@@ -0,0 +1,277 @@
+			"Instance with instance-ID doesn't exist",
+			fmt.Errorf("Instance with instance-ID doesn't exist"),
+		)
+	} else if *input.DryRun {
+		// If it is a dry run
+		return nil, awserr.New(
+			"DryRunOperation",
+			"This is a dryRun call",
+			fmt.Errorf("This is a dry run call"),
+		)
+	}
+
+	var desiredInstance ec2.Instance
+	found := false
+	newInstanceList := make([]ec2.Instance, 0)
+
+	for _, instanceID := range input.InstanceIds {
+		for _, instance := range *ms.FakeInstances {
+			if *instance.InstanceId == *instanceID {
+				// Do not append InstanceID, there by removing it
+				found = true
+				desiredInstance = instance
+			} else {
+				newInstanceList = append(newInstanceList, instance)
+			}
+		}
+	}
+	ms.FakeInstances = &newInstanceList
+
+	if !found {
+		return nil, fmt.Errorf("Couldn't find instance with given instance-ID %s", *input.InstanceIds[0])
+	}
+
+	return &ec2.TerminateInstancesOutput{
+		TerminatingInstances: []*ec2.InstanceStateChange{
+			&ec2.InstanceStateChange{
+				PreviousState: desiredInstance.State,
+				InstanceId:    input.InstanceIds[0],
+				CurrentState: &ec2.InstanceState{
+					Code: aws.Int64(32),
+					Name: aws.String("shutting-down"),
+				},
+			},
+		},
+	}, nil
+}
+
+// StopInstances implements a mock stop instance method
+func (ms *MockEC2Client) StopInstances(input *ec2.StopInstancesInput) (*ec2.StopInstancesOutput, error) {
+
+	if *input.InstanceIds[0] == InstanceStopError {
+		return nil, fmt.Errorf("Stopping of instance errored out")
+	} else if *input.InstanceIds[0] == InstanceDoesntExistError {
+		// If instance with instance ID doesn't exist
+		return nil, awserr.New(
+			ec2.UnsuccessfulInstanceCreditSpecificationErrorCodeInvalidInstanceIdMalformed,
+			"Instance with instance-ID doesn't exist",
+			fmt.Errorf("Instance with instance-ID doesn't exist"),
+		)
+	} else if *input.DryRun {
+		// If it is a dry run
+		return nil, awserr.New(
+			"DryRunOperation",
+			"This is a dryRun call",
+			fmt.Errorf("This is a dry run call"),
+		)
+	}
+
+	var desiredInstance ec2.Instance
+	found := false
+
+	for _, instanceID := range input.InstanceIds {
+		for _, instance := range *ms.FakeInstances {
+			if *instance.InstanceId == *instanceID {
+				// Do not append InstanceID, there by removing it
+				found = true
+				desiredInstance = instance
+			} else {
+			}
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("Couldn't find any instance matching requirement")
+	}
+
+	return &ec2.StopInstancesOutput{
+		StoppingInstances: []*ec2.InstanceStateChange{
+			&ec2.InstanceStateChange{
+				PreviousState: desiredInstance.State,
+				InstanceId:    input.InstanceIds[0],
+				CurrentState: &ec2.InstanceState{
+					Code: aws.Int64(64),
+					Name: aws.String("stopping"),
+				},
+			},
+		},
+	}, nil
+}
+
+// deepCopyTagList copies inTags list to outTags
+func deepCopyTagList(inTags []*ec2.Tag) []*ec2.Tag {
+	var outTags []*ec2.Tag
+
+	for _, tagPtr := range inTags {
+		tag := *tagPtr
```

Comment:
```suggestion
		tag := &ec2.Tag{}
		if tagPtr.Key != nil {
		  tag.Key = *tagPtr.Key
		}
		if tagPtr.Value != nil {
		  tag.Value = *tagPtr.Value
		}
```
Just `tag := *tagPtr` is just shallow copy of the element in the slice.

Reply:
I had the same doubt in my mind. Hence wanted your review. Thanks :)